tls/s2n_prf.c

```diff
@@ -382,17 +382,27 @@ int s2n_prf_key_expansion(struct s2n_connection *conn)
 
     /* Check that we have a valid MAC and key size */
     int mac_size;
-    GUARD((mac_size = s2n_hmac_digest_size(hmac_alg)));
+    if (conn->pending.cipher_suite->cipher->type == S2N_COMP) {
+        mac_size = conn->pending.cipher_suite->cipher->io.comp.mac_key_size;
+    } else {
+        GUARD((mac_size = s2n_hmac_digest_size(hmac_alg)));
+    }
 
     /* Seed the client MAC */
-    uint8_t *client_write_mac_key = s2n_stuffer_raw_read(&key_material, mac_size);
-    notnull_check(client_write_mac_key);
-    GUARD(s2n_hmac_init(&conn->pending.client_record_mac, hmac_alg, client_write_mac_key, mac_size));
+    uint8_t *client_mac_write_key = s2n_stuffer_raw_read(&key_material, mac_size);
+    notnull_check(client_mac_write_key);
+    GUARD(s2n_hmac_init(&conn->pending.client_record_mac, hmac_alg, client_mac_write_key, mac_size));
 
     /* Seed the server MAC */
-    uint8_t *server_write_mac_key = s2n_stuffer_raw_read(&key_material, mac_size);
-    notnull_check(server_write_mac_key);
-    GUARD(s2n_hmac_init(&conn->pending.server_record_mac, hmac_alg, server_write_mac_key, mac_size));
+    uint8_t *server_mac_write_key = s2n_stuffer_raw_read(&key_material, mac_size);
+    notnull_check(server_mac_write_key);
+    GUARD(s2n_hmac_init(&conn->pending.server_record_mac, hmac_alg, server_mac_write_key, mac_size));
+
+    /* Composite CBC does MAC inside the cipher, so pass them the MAC key */
+    if (conn->pending.cipher_suite->cipher->type == S2N_COMP) {
+        GUARD(conn->pending.cipher_suite->cipher->io.comp.get_mac_write_key(&conn->pending.server_key, server_mac_write_key, mac_size));
+        GUARD(conn->pending.cipher_suite->cipher->io.comp.get_mac_write_key(&conn->pending.client_key, client_mac_write_key, mac_size));
```

Comment:
Maybe something like this: 46dfc1178? Make the server_key and client_key fully formed.